CRM/Contact/BAO/Contact.php

```diff
@@ -2629,6 +2629,115 @@ public static function getCountComponent($component, $contactId, $tableName = NU
     }
   }
 
+  /**
+   * Check whether greetings for this contact contain tokens for the given
+   * custom field.
+   *
+   * @param object $contact Contact object.
+   * @param int $customFieldId System ID of the custom field.
+   * @return bool
+   */
+  public static function checkGreetingsUseCustomFieldToken($contact, $customFieldId) {
+    $greetingStrings = self::getGreetingStrings($contact, FALSE);
+    foreach ($greetingStrings as $greetingString) {
+      if (strpos($greetingString, ".custom_{$customFieldId}}") !== FALSE) {
+        return TRUE;
+      }
+    }
+    return FALSE;
+  }
+
+  /**
+   * Get an array of template strings for greeting fields.
+   *
+   * @param object $contact Contact object.
+   * @param bool $useDefaults If TRUE, use system default greeting values;
+   *    otherwise, use values for this specific contact.
+   * @return Array of greeting strings, with keys: 'email', 'postal', 'addressee'.
+   */
+  public static function getGreetingStrings($contact, $useDefaults = FALSE) {
+    Civi::$statics[__CLASS__]['greeting_strings'] = array();
```

Comment:
I think any performance benefit of caching the strings by contact is outweighed by the risk that it could become stale in an update scenario.

What about re-adding a function to get the defaults & then in getGreetingStrings you simply have 

return array_merge(self::getGreetingDefaultsByContactType($params['contact_type'], $contact);


  /**
   * Get the default details for selected type
   *
   * @param array
   */
  public static function getDefaultGreetingsForContactType($contactTypeToRetrieve) {

    if (!isset(Civi::$statics[__CLASS__])) {
      Civi::$statics[__CLASS__]['greetings'] = array();

      $optionGroups = civicrm_api3('OptionGroup', 'get', array(
          'name' => array(
            'IN' => array(
              'email_greeting',
              'postal_greeting',
              'addressee'
            )
          )
        )
      );

      $optionValues = civicrm_api3('OptionValue', 'get', array(
        'option_group_id' => array('IN' => array_keys($optionGroups['values'])),
        'is_default' => 1,
      ));

      $contactTypes = civicrm_api3('ContactType', 'get', array('is_reserved' => 1));
      foreach ($contactTypes as $contactTypeID => $contactType) {
        foreach ($optionValues as $id => $optionValue) {
          if ($optionValue['filter'] === $contactTypeID) {
            $keyName = ($optionValue['name'] === 'addressee' ? 'addressee' : $optionValue['name'] . '_id');
            Civi::$statics[__CLASS__]['greetings'][$contactType][$keyName] = $optionValue['value'];
          }
        }
      }
    }
    return Civi::$statics[__CLASS__]['greetings'][$contactTypeToRetrieve];
  }